### PR TITLE
 Optimize and Cleanup CScript::FindAndDelete

### PR DESCRIPTION
**Pull request by pstratem**
This PR improves the worst-case behavior of CScript::FindAndDelete.

I have emphasized the obvious correctness of the algorithm in this commit.

Additionally I have done extensive fuzzing to ensure the result is identical to the current implementation.

### CODE REVIEW

**Comment by kazcw** [comment]
Is this really an optimization? It increases copying in the non-multisig case, which is vastly more common...

**Comment by TheBlueMatt** [comment]
I highly prefer this to #7884...the original is kinda hard to read, but #7884 is impossible to reasonably convince yourself works, even if it probably does.
I benchmarked syncing on tmpfs with the final call to VerifySignature in CheckSig() disabled on this and #7884 and couldn't detect any difference in sync time through about 280k (just happened to be some blk* flies I had lying around went until there).

File: src/script/script.h [hunk 1]
```diff
@@ -570,17 +570,25 @@ class CScript : public CScriptBase
...
-                pc = erase(pc, pc + b.size());
+                pc = pc + b.size();
                 ++nFound;
             }
+            pc2 = pc;
         }
         while (GetOp(pc, opcode));
+
+        result.insert(result.end(), pc2, end());
+        *this = result;
```
Comment (TheBlueMatt) [inline src/script/script.h]: You should gate this and the last insert on an if(nFound)

**Comment by dcousens** [comment]
utACK 6443169,  but I'd prefer benchmarks before merge.

**Comment by sipa** [comment]
@kazwc @dcousens It is only intended to improve the worst case behaviour, and @TheBlueMatt already benchmarked that it does not measurably affect the average case.

@pstratem If, after includong Matt's suggestion, you drop the result.reserve() call, I think this will run with zero heap effect for non-match situations.

**Comment by pstratem** [comment]
@sipa indeed you're right I've made that change

**Comment by dcousens** [comment]
@sipa if we can't measure this is an improvement,  isn't this just all speculation?

**Comment by dcousens** [comment]
utACK da1d7775,  looks good to merge

**Comment by laanwj** [comment]
utACK d1d7775

**Comment by sipa** [comment]
@dcousens it gives a very measurable improvement in worst case situations.
Just not in the average case you'll observe with a random benchmark.

**Comment by sipa** [comment]
utACK d1d7775587473410a107e7079616b9ecaae8dd06

File: src/test/script_tests.cpp [hunk 2]
```diff
@@ -1051,4 +1051,121 @@ BOOST_AUTO_TEST_CASE(script_GetScriptAsm)
...
+    BOOST_CHECK(s == expect);
+
+    s = ScriptFromHex("0003feed");
+    d = ScriptFromHex("00");
+    expect = ScriptFromHex("03feed");
+    BOOST_CHECK_EQUAL(s.FindAndDelete(d), 1);
+    BOOST_CHECK(s == expect);
+}
+
 BOOST_AUTO_TEST_SUITE_END()
```
Comment (kellwin13) [inline src/test/script_tests.cpp]: kellwin

**Pull request merged**